js/src/path-test.js

```diff
@@ -8,100 +8,277 @@ import {assert} from 'chai';
+    await resolvesTo(s, l, l);
+    await resolvesTo(s, lr, lr);
+
+    // Lists cannot be addressed by hashes, obviously.
+    await resolvesTo(l, null, i);
+  });
+
+  test('hash index of singleton collection', async () => {
+    // This test is to make sure we don't accidentally return the element of a singleton map.
+    const resolvesToNull = async (col: Value, v: Value) => {
+      await assertResolvesTo(null, col, hashIdx(v));
+    };
+
+    await resolvesToNull(new Map([[true, true]]), true);
+    await resolvesToNull(new Set([true]), true);
   });
 
   test('multi', async () => {
-    const m1 = new Map([['a', 'foo'], ['b','bar'], ['c', 'car']]);
-    const m2 = new Map([['d', 'dar'], [false, 'earth']]);
+    const m1 = new Map([
+      ['a', 'foo'],
+      ['b', 'bar'],
+      ['c', 'car'],
+    ]);
+
+    const m2 = new Map([
+      ['d', 'dar'],
+      [false, 'earth'],
+      [m1, 'fire'],
+    ]);
+
     const l = new List([m1, m2]);
+
     const s = newStruct('', {
-      foo: l,
+      'foo': l,
     });
 
-    await assertPathEqual(l, s, new Path().addField('foo'));
-    await assertPathEqual(m1, s, new Path().addField('foo').addIndex(0));
-    await assertPathEqual('foo', s, new Path().addField('foo').addIndex(0).addIndex('a'));
-    await assertPathEqual('bar', s, new Path().addField('foo').addIndex(0).addIndex('b'));
-    await assertPathEqual('car', s, new Path().addField('foo').addIndex(0).addIndex('c'));
-    await assertPathEqual(undefined, s, new Path().addField('foo').addIndex(0).addIndex('x'));
-    await assertPathEqual(undefined, s, new Path().addField('foo').addIndex(2).addIndex('c'));
-    await assertPathEqual(undefined, s, new Path().addField('notHere').addIndex(2).addIndex('c'));
-    await assertPathEqual(m2, s, new Path().addField('foo').addIndex(1));
-    await assertPathEqual('dar', s, new Path().addField('foo').addIndex(1).addIndex('d'));
-    await assertPathEqual('earth', s, new Path().addField('foo').addIndex(1).addIndex(false));
+    await assertResolvesTo(l, s, '.foo');
+    await assertResolvesTo(m1, s, '.foo[0]');
+    await assertResolvesTo('foo', s, '.foo[0]["a"]');
+    await assertResolvesTo('bar', s, '.foo[0]["b"]');
+    await assertResolvesTo('car', s, '.foo[0]["c"]');
+    await assertResolvesTo(null, s, '.foo[0]["x"]');
+    await assertResolvesTo(null, s, '.foo[2]["c"]');
+    await assertResolvesTo(null, s, '.notHere[0]["c"]');
+    await assertResolvesTo(m2, s, '.foo[1]');
+    await assertResolvesTo('dar', s, '.foo[1]["d"]');
+    await assertResolvesTo('earth', s, '.foo[1][false]');
+    await assertResolvesTo('fire', s, `.foo[1]${hashIdx(m1)}`);
+    await assertResolvesTo(m1, s, `.foo[1]${hashIdx(m1)}@key`);
+    await assertResolvesTo('car', s, `.foo[1]${hashIdx(m1)}@key["c"]`);
   });
 
-  function assertToString(expect: string, path: Path) {
-    assert.strictEqual(expect, path.toString());
-  }
+  test('parse success', () => {
+    const test = (s: string) => {
```

Comment:
Can you call this something else than test? It is shadowing the outer one which might be a bit confusing.